src/models/transfers.js

```diff
@@ -231,24 +233,26 @@ function * fulfillTransfer (transferId, fulfillment) {
 
     // We don't know if notary is executing or rejecting transfer
     const isValidExecution = Boolean(transfer.execution_condition &&
-      Condition.testFulfillment(transfer.execution_condition, fulfillment))
+      transfer.execution_condition.type === fulfillment.getData().condition_fulfillment.type &&
+      Condition.testFulfillment(transfer.execution_condition, fulfillment.getData().condition_fulfillment))
     const isValidCancellation = Boolean(transfer.cancellation_condition &&
-      Condition.testFulfillment(transfer.cancellation_condition, fulfillment))
+      transfer.cancellation_condition.type === fulfillment.getData().condition_fulfillment.type &&
+      Condition.testFulfillment(transfer.cancellation_condition, fulfillment.getData().condition_fulfillment))
 
     if (isValidCancellation === isValidExecution) {
       throw new UnmetConditionError('ConditionFulfillment failed')
     }
 
-    const canExecute = transfer.state === 'prepared'
-    const canCancel = transfer.state === 'proposed' || transfer.state === 'prepared'
+    const canExecute = transfer.state === 'prepared' || transfer.state === 'executed'
```

Comment:
I don't see how to make PUT idempotent, since it will throw an error if the transfer is not yet created (did we agree this is not idempotent behavior?). I think the behavior I'd like to support is more consistent with POST rather than PUT. But it is an API breaking change. What do you think @naoitoi?